### PR TITLE
hot reloading

### PR DESCRIPTION
**Pull request by adirar111**
we HAD to rename the `main` directory to `content` because the hot reload package assumes that's what it's called (and i was getting gnarly errors)

(probably unmergable rn because of the conflicts)

the cool thing is now we can make changes in `content`/`background` and it automatically reloads the page and extension

it doesn't work for `options`, so any changes made there need to be manually refreshed (or i guess you could coerce a refresh by changing something in content/background if you're lazy)

bruh 2

### CODE REVIEW

File: webpack.config.ts [hunk 1]
```diff
@@ -1,32 +1,56 @@
...
+		client: {
+			host: "localhost",
+		},
+	},
 	entry: {
-		'bundle': './ts/main/index.ts',
-		'options': './ts/options/index.ts',
-		'background': './ts/background/delegator.ts',
+		options: path.join(tsSrcDir, "options"),
+		bundle: getEntry("content"),
```
Comment (braxtonhall) [inline webpack.config.ts:33]: wait why can't this be `"main"`?

**Comment by braxtonhall** [comment]
OH options doesn't need a hot reload. It's just up to date I _think_

File: webpack.config.ts (hunk 1 above)
Comment (adirar111) [reply webpack.config.ts:33]:
```ts
[watch:*src] ERROR in bundle
[watch:*src] Module not found: Error: Can't resolve 'mv3-hot-reload/main' in '/Users/aymendirar/Developer/library-thing'
[watch:*src] resolve 'mv3-hot-reload/main' in '/Users/aymendirar/Developer/library-thing'
[watch:*src]   Parsed request is a module
[watch:*src]   using description file: /Users/aymendirar/Developer/library-thing/package.json (relative path: .)
[watch:*src]     Field 'browser' doesn't contain a valid alias configuration
[watch:*src]     resolve as module
[watch:*src]       looking for modules in /Users/aymendirar/Developer/library-thing/node_modules
[watch:*src]         existing directory /Users/aymendirar/Developer/library-thing/node_modules/mv3-hot-reload
[watch:*src]           using description file: /Users/aymendirar/Developer/library-thing/node_modules/mv3-hot-reload/package.json (relative path: .)
[watch:*src]             using description file: /Users/aymendirar/Developer/library-thing/node_modules/mv3-hot-reload/package.json (relative path: ./main)
[watch:*src]               no extension
[watch:*src]                 Field 'browser' doesn't contain a valid alias configuration
[watch:*src]                 /Users/aymendirar/Developer/library-thing/node_modules/mv3-hot-reload/main doesn't exist
[watch:*src]               .tsx
[watch:*src]                 Field 'browser' doesn't contain a valid alias configuration
[watch:*src]                 /Users/aymendirar/Developer/library-thing/node_modules/mv3-hot-reload/main.tsx doesn't exist
[watch:*src]               .ts
[watch:*src]                 Field 'browser' doesn't contain a valid alias configuration
[watch:*src]                 /Users/aymendirar/Developer/library-thing/node_modules/mv3-hot-reload/main.ts doesn't exist
[watch:*src]               .js
[watch:*src]                 Field 'browser' doesn't contain a valid alias configuration
[watch:*src]                 /Users/aymendirar/Developer/library-thing/node_modules/mv3-hot-reload/main.js doesn't exist
[watch:*src]               as directory
[watch:*src]                 /Users/aymendirar/Developer/library-thing/node_modules/mv3-hot-reload/main doesn't exist
[watch:*src]       /Users/aymendirar/Developer/node_modules doesn't exist or is not a directory
[watch:*src]       /Users/aymendirar/node_modules doesn't exist or is not a directory
[watch:*src]       /Users/node_modules doesn't exist or is not a directory
[watch:*src]       /node_modules doesn't exist or is not a directory
[watch:*src] 
[watch:*src] webpack 5.74.0 compiled with 1 error in 3442 ms
```

it hates being called `main` -- i think they assume it's called `common` to do whatever they're doing and renaming it fixed it. it also kinda coincides with what is -- content scripts vs background stuff

File: package.json [hunk 2]
```diff
@@ -28,7 +30,9 @@
 		"webpack-cli": "^4.10.0"
 	},
 	"scripts": {
-		"watch": "webpack --watch --mode development",
+		"watch:src": "webpack --watch --mode development",
+		"watch:dist": "mv3-hot-reload",
+		"dev": "concurrently yarn:watch:*",
```
Comment (braxtonhall) [inline package.json:35]: what the heck does the thing also not allow the npm script to be named `"watch"`???

Comment (adirar111) [reply package.json:35]: nah i just stole this from them but i'm bringing `watch` back so we don't have to think about running a diff command lol

Comment (braxtonhall) [reply package.json]: i mean idc either way, but does it not work with `"watch"`, `"dev:webpack"`, and `"dev:mv3"`?

File: webpack.config.ts [hunk 3]
```diff
@@ -1,32 +1,41 @@
...
-	devtool: options.mode !== 'production' ? 'source-map' : undefined,
-	context: path.join(__dirname, '/extension'),
+import config from "./mv3-hot-reload.config";
+
+const isDev = (options: WebpackOptionsNormalized) => options.mode !== "production";
+
+const srcDir = path.join(__dirname, "extension");
+const tsSrcDir = path.join(srcDir, "ts");
+const getEntry = (name: string) => {
+	return [path.join(tsSrcDir, name), ...(isDev ? [`mv3-hot-reload/${name}`] : [])];
```
Comment (braxtonhall) [inline webpack.config.ts:11]: `isDev` function is not being applied here

File: package.json (hunk 2 above)
Comment (adirar111) [reply package.json]: oh oop i think it does

File: webpack.config.ts (hunk 3 above)
Comment (adirar111) [reply webpack.config.ts:11]: oopsie

**Comment by adirar111** [comment]
closing in favor of #162

**Review by braxtonhall — APPROVED** [review]
this one? we can squash this one

**Pull request merged**